### ISSUE TITLE
regression 1.50: Borrow checker reports function param as not living long enough

### ISSUE THREAD

**Issue by rylev** [rust-lang/rust#80949]
This might not actually be a regression an instead the code may have previously erroneously compiled. I cannot find an an open issue obviously related to this though.

As you can see [here](https://crater-reports.s3.amazonaws.com/beta-1.50-1/beta-2021-01-01/reg/euca-0.1.0/log.txt) this code does not compile with a "borrowed value does not live long enough" where it did previously. This comes from a call to `diff::diff` which has [this signature](https://docs.rs/euca/0.1.0/euca/diff/fn.diff.html). 

The error message states that `n` may be dropped while the borrow in the next param `state` is still active. This fails to compile on both 1.50 and 1.49 if the param `state` is move to a local variable. 

@rustbot modify labels: +regression-from-stable-to-beta

**Comment by apiraino** [comment]
Let's try to reduce it @rustbot ping cleanup

**Comment by hellow554** [comment]
Still with dependency, but simplified code

```
use euca::diff;
use euca::dom::Dom;
use euca::vdom::DomIter;

#[test]
fn from_empty() {
    let new: Dom<(), (), &()> = Dom::elem("div");

    let n = new.dom_iter();
    let mut storage = vec![];
    let patch_set = diff::diff(std::iter::empty(), n, &mut storage);
}
```

let's see what I can get from there

**Comment by hellow554** [comment]
searched nightlies: from nightly-2020-12-01 to nightly-2021-01-06
regressed nightly: nightly-2020-12-06
searched commits: from https://github.com/rust-lang/rust/commit/3ff10e74a74ed093fcabac1de27fe1cd65bbbb4a to https://github.com/rust-lang/rust/commit/e792288df31636ca28108516c63a00ce4267063a
regressed commit: https://github.com/rust-lang/rust/commit/9122b769c8306b1cb3c8d1f96fca3ea3e208e22e

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.0</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc 2021-01-06 --without-cargo --access github --regress error
```
</details>

The code I tested:


```
pub struct Dom<Message>(Message);
impl<Message> Dom<Message> {
    pub fn new() -> Self {
        todo!()
    }
    fn dom_iter(&self) -> Box<dyn Iterator<Item = &Message>> {
        todo!()
    }
}

pub fn diff<'a, Message, O, N, S>(_: O, _: N, _: S) -> &'a Message
where
    O: IntoIterator<Item = &'a Message>,
    N: IntoIterator<Item = &'a Message>,
    S: IntoIterator<Item = &'a Message>,
{
    todo!()
}

fn main() {
    let new: Dom<()> = Dom::new();

    let n = new.dom_iter();
    let storage = vec![];
    diff(std::iter::empty(), n, &storage);
}
```

**Comment by matthewjasper** [comment]
So #78373 resulted in MIR having one fewer block, which is making borrowck be less precise with how long the borrow lasts. There are a few possible fixes here for me to look at.

**Comment by apiraino** [comment]
Assigning `P-high` as discussed as part of the [Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2380949.20regression.201.2E50.3A.20Borrow.20checker.20reports.20function.20par.E2.80.A6) and removing `I-prioritize`.

**Comment by Mark-Simulacrum** [comment]
@rylev how common was this bug in the crater report? just the one crate?

**Comment by rylev** [comment]
Yes the [linked crate](https://crater-reports.s3.amazonaws.com/beta-1.50-1/beta-2021-01-01/reg/euca-0.1.0/log.txt) was the only crate to show signs of this issue in the crater run.

**Comment by matthewjasper** [comment]
@nikomatsakis @pnkfelix @nagisa wanted a longer explanation of what's going on here.

When we type check the MIR we find that:
* The lifetime of the borrow at bb3[3] must outlive the lifetime in the type of `_8`
* the lifetimes in the types of `_8` and `_7` are the same because of the signature of `diff`
* The lifetime in the type of `_7` is the same as the lifetime in the type of `_1` (from the assignment at bb3[1])

So the lifetime of the borrow contains all points where `_8`, `_7` and `_1` are live and initialized, which I have marked in green. The borrow therefore lasts only for bb3[3] and bb3[4] (borrows only start at the statement with the borrow rvalue).

![old-mir](https://user-images.githubusercontent.com/20113453/104633974-dfd10500-5697-11eb-8bd6-695f91faaca6.png)

In the new MIR the drop of `_7` is gone, so there's no longer a gap in the lifetime between bb3 and the drop of `_4`. This results in the borrow of `_4` including bb7 and bb8, which conflicts with the drop of `_4` in bb7.

![new-mir](https://user-images.githubusercontent.com/20113453/104633973-df386e80-5697-11eb-876d-877141611ec8.png)

**Comment by pnkfelix** [comment]
discussed at T-compiler meeting. We want to resolve this issue, and plan to revert PR #78373 as a short-term resolution. Longer term will be to re-land PR #78373 with hopefully some further modifications so that we don't cause regressions like this in the process.

**Comment by pnkfelix** [comment]
While setting up my system for the revert, I narrowed the MCVE down a bit further. I don't think I illuminated much more in the process, though I do find it curious that I wasn't able to easily find a way to take the `Box<dyn Trait>` out of the equation for reproducing the bug:

[playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=3502a305989c8aaa4f8e6f051c7d6957)

```rust
trait Trait { type Item; }

impl<'a, X> Trait for &'a Vec<X> {
    type Item = &'a X;
}

impl<X> Trait for Box<dyn Trait<Item = X>> {
    type Item = X;
}

fn make_dyn_trait(_: &()) -> Box<dyn Trait<Item = &()>> {
    todo!()
}

fn diff<'a, M, N, S>(_: N, _: S)
where
    M: 'a,
    N: Trait<Item = &'a M>,
    S: Trait<Item = &'a M>,
{
    todo!()
}

fn may_panic<X>(_: X) { }

fn main() {
    let dyn_trait = make_dyn_trait(&());
    let storage = vec![()];
    let _x = may_panic(());
    let storage_ref = &storage;
    diff(dyn_trait, storage_ref);
}
```

**Comment by mzabaluev** [comment]
Here is a minified example of a regression I have run into, not sure if it's the same issue:

```rust
use std::error::Error;
use std::iter;

fn walk_source_chain(error: &impl Error) {
    for e in iter::successors(error.source(), |e| e.source()) {
        println!("{}", e);
    }
}
```

Changing the closure parameter pattern to match `&e` fixes the borrow checker error in nightly.

**Comment by camelid** [comment]
Hmm, that example compiles on beta though.

**Comment by camelid** [comment]
I believe that's an unrelated issue. Opened #81460 to track it. Thanks for reporting it!

**Comment by pietroalbini** [comment]
Fix backported to 1.50 and landed in 1.51, closing this.
